purple-pill/js/controllers/fontexportController.js

```diff
@@ -32,140 +112,107 @@ function($scope, $http, sharedScope, $timeout) {
-              ;
-            io[index] = exportObjects[1];
-            return gen;
-        }
-
-        var current_instance = 0
-          , total_instances = instances_for_export.length
-          , glyphs_phase_percentage = 80 //An estimated 80% of the time is taken generating the files. The other 20% is spent zipping the results.
-          , UI_UPDATE_TIMESLICE = 50 // msecs
-          ;
-          
-        function setProgress(width, text) {
-            $("#progressbar").animate({"opacity": 1, "width": width + "%"}, /*duration:*/ UI_UPDATE_TIMESLICE);
-            if (text)
-                $("#progresslabel").html(text);
-        }
-
         function setDownloadBlobLink(text, blob, filename) {
-            $("#progressbar").animate({"width": "100%", "opacity": 1}, /*duration:*/ UI_UPDATE_TIMESLICE);
-            $("#progresslabel").html("");
-            $("#blob_download").css("display", "block");
-            $("#blob_download").children("a").html(text).click(function(){
+            var download = $("#blob_download");
+            progress.complete();
+
+            download.css("display", "block");
+            download.children("a").html(text).click(function(){
                 $scope.data.stateless.saveAs(blob, filename);
-                resetProgressBar();
-                delete bundle_data;
+                progress.reset();
+                download.css("display", "none").children("a").unbind("click");
+                delete bundleData;
             });
         }
-        
-        function resetProgressBar() {
-            $("#progressbar").animate({"opacity": 0, "width": 0}, /*duration:*/ 0); // (that means "do it immediately!")
-            $("#progresslabel").html("");
-            $("#blob_download").css("display", "none").children("a").unbind("click");
+
+        function exportingGlyphMessage (it, instanceIndex){
+            var msg
+              , currentGlyph = it.value['current_glyph'] + 1 //humans start counting from 1.
+              , totalGlyphs = it.value['total_glyphs']
+              , glyphId = it.value['glyph_id']
+              ;
+            msg = totalInstances + " instances to export."
+            msg += "Calculating glyph '" + glyphId + "' (" + currentGlyph + " of " + totalGlyphs + ")"
+            msg += " of instance #" + (instanceIndex+1);
+            return msg;
+        }
+
+        function calculateGlyphsProgress(it, instanceIndex, totalInstances){
+            var percentage
+              , currentGlyph = it.value['current_glyph']
+              , totalGlyphs = it.value['total_glyphs']
+              ;
+            percentage = (instanceIndex + (currentGlyph / totalGlyphs)) / totalInstances;
+            return percentage;
         }
 
-        function exportFont_compute_glyphs(){
+        var index = 0
+          , totalInstances = exportObjects.length
+          , obj = exportObjects.pop()
+          ;
+        function exportFontComputeGlyphs(){
```

Comment:
Thinking about this. If you would split this into two functions, one as a starter function and one as the 'loop', you could niceley use promises:

```js
// the last two arguments are the `executor` interface of 
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
// note, this is your original function, I prefixed it with a _ underscore to make sure it's not called directly
function _exportFont_compute_glyphs(exportObjects, totalInstances, bundleFolder, resolve, reject) {
    if (!exportObjects.length) {
        resolve(true);
        return;
    }
    var obj = exportObjects[exportObjects.length-1]
      , data, name
      , it = obj.generator.next()
      ;
    if (!it.done) {
        // NOTE: run `exportingGlyphMessage` and `calculateGlyphsProgress` in progress.set
        // NOTE: that we don't pass `it`, but just `it.value`
        // NOTE: index = totalInstances - exportObjects.length
        progress.set(it.value, totalInstances - exportObjects.length, totalInstances);
    }
    else {
        exportObjects.pop();
        obj.pruneGenerator(); // don't know if we need this. obj is void in a second anyways
        if (obj.fileFormat == "UFO") {
            data = $scope.data.stateful.project.getZipFromIo(false, obj.io, obj.getFileName(), "uint8array");
            name = obj.getFileName()+".zip";
        }
        else { /* obj.fileFormat == "OTF" */
            data = obj.io.readFile(false, obj.getFileName());
            name = obj.getFileName();
        }
         bundleFolder.file(name, data, {binary:true});
    }
    $timeout(exportFontComputeGlyphs.bind(null, exportObjects, totalInstances, bundle, bundleFolder, resolve, reject)
                , UI_UPDATE_TIMESLICE);
}

// this is the initiating method

function exportFont_compute_glyphs(exportObjects, bundleFolderName) {
    var bundle = new $scope.data.stateless.JSZip()
      , bundleFolder = bundle.folder(bundleFolderName)
      , totalInstances = exportObjects.length
      ;
    // note how the first three args are bound, new Promise will call the bound function with
    // the two missing args `resolve` and `reject`
    return new Promise(
        _exportFont_compute_glyphs.bind(null, exportObjects, totalInstances, bundleFolder)
    ).then(function(){
        return bundle.generate({type:"blob"});
    });
}


// to execute the export you now do:
function finalizeExport(bundleData) {
    setDownloadBlobLink(bundleFileName, bundleData, bundleFileName);
    exportIsRunning = false;
}
exportFont_compute_glyphs(exportObjects).then(finalizeExport);
```